Source/Core/VideoCommon/TextureConversionShader.cpp

```diff
@@ -2,1397 +2,1421 @@
+             "  float2 clamp_tb;\n"
+             "  float3 filter_coefficients;\n"
+             "};\n");
+      if (g_ActiveConfig.backend_info.bSupportsGeometryShaders)
+      {
+        Append("VARYING_LOCATION(0) in VertexData {\n"
+               "  float3 v_tex0;\n"
+               "};\n");
+      }
+      else
+      {
+        Append("VARYING_LOCATION(0) in float3 v_tex0;\n");
+      }
+      Append("SAMPLER_BINDING(0) uniform sampler2DArray samp0;\n"
+             "FRAGMENT_OUTPUT_LOCATION(0) out float4 ocol0;\n");
+    }
+    else  // D3D
+    {
+      Append("cbuffer PSBlock : register(b0) {\n"
+             "  int4 position;\n"
+             "  float y_scale;\n"
+             "  float gamma_rcp;\n"
+             "  float2 clamp_tb;\n"
+             "  float3 filter_coefficients;\n"
+             "};\n"
+             "sampler samp0 : register(s0);\n"
+             "Texture2DArray Tex0 : register(t0);\n");
+    }
 
-static void WriteZ16LEncoder(char*& p, APIType ApiType, const EFBCopyParams& params)
-{
-  WriteSwizzler(p, params, EFBCopyFormat::GB8, ApiType);
+    // D3D does not have roundEven(), only round(), which is specified "to the nearest integer".
+    // This differs from the roundEven() behavior, but to get consistency across drivers in OpenGL
+    // we need to use roundEven().
+    if (api_type == APIType::D3D)
+      Append("#define roundEven(x) round(x)\n");
+
+    // Alpha channel in the copy is set to 1 the EFB format does not have an alpha channel.
+    Append("float4 RGBA8ToRGB8(float4 src)\n"
+           "{\n"
+           "  return float4(src.xyz, 1.0);\n"
+           "}\n"
+           "float4 RGBA8ToRGBA6(float4 src)\n"
+           "{\n"
+           "  int4 val = int4(roundEven(src * 255.0)) >> 2;\n"
+           "  return float4(val) / 63.0;\n"
+           "}\n"
+           "float4 RGBA8ToRGB565(float4 src)\n"
+           "{\n"
+           "  int4 val = int4(roundEven(src * 255.0));\n"
+           "  val = int4(val.r >> 3, val.g >> 2, val.b >> 3, 1);\n"
+           "  return float4(val) / float4(31.0, 63.0, 31.0, 1.0);\n"
+           "}\n");
+  }
 
-  WRITE(p, "  float depth;\n");
-  WRITE(p, "  float3 expanded;\n");
+  void WriteSampleFunction(const EFBCopyParams& params, APIType api_type)
+  {
+    const auto WriteSampleOp = [api_type, &params, this](int y_offset) {
+      if (!params.depth)
+      {
+        switch (params.efb_format)
+        {
+        case PEControl::RGB8_Z24:
+          Append("RGBA8ToRGB8(");
+          break;
+        case PEControl::RGBA6_Z24:
+          Append("RGBA8ToRGBA6(");
+          break;
+        case PEControl::RGB565_Z16:
+          Append("RGBA8ToRGB565(");
+          break;
+        default:
+          Append("(");
+          break;
+        }
+      }
+      else
+      {
+        // Handle D3D depth inversion.
+        if (!g_ActiveConfig.backend_info.bSupportsReversedDepthRange)
+          Append("1.0 - (");
+        else
+          Append("()");
```

Comment:
This looks weird. Everything else only writes an opening parenthesis, but no closing ones.
Then again, I have no idea about shader code 🤷‍♂